src/axom/quest/DistributedClosestPoint.hpp

```diff
@@ -442,186 +522,238 @@ class DistributedClosestPointImpl
+          {
+            axom::copy(qmcpr.data_ptr(),
+                       xferNode.fetch_existing("cp_rank").data_ptr(),
+                       qPtCount * sizeof(axom::IndexType));
+          }
+          if(xferNode.fetch_existing("cp_index").data_ptr() != qmcpi.data_ptr())
+          {
+            axom::copy(qmcpi.data_ptr(),
+                       xferNode.fetch_existing("cp_index").data_ptr(),
+                       qPtCount * sizeof(axom::IndexType));
+          }
+          if(xferNode.fetch_existing("closest_point").data_ptr() !=
+             qmcpcp.data_ptr())
+          {
+            axom::copy(qmcpcp.data_ptr(),
+                       xferNode.fetch_existing("closest_point").data_ptr(),
+                       qPtCount * sizeof(PointType));
+          }
 
-        // copy data to mesh_node from proc_node
-        const int npts = proc_node["npts"].value();
-        axom::copy(xfer_node["cp_rank"].data_ptr(),
-                   proc_node["cp_rank"].data_ptr(),
-                   npts * sizeof(axom::IndexType));
-        axom::copy(xfer_node["cp_index"].data_ptr(),
-                   proc_node["cp_index"].data_ptr(),
-                   npts * sizeof(axom::IndexType));
-        axom::copy(xfer_node["closest_point"].data_ptr(),
-                   proc_node["closest_point"].data_ptr(),
-                   npts * sizeof(PointType));
-
-        if(m_isVerbose)
-        {
-          dumpNode(mesh_node,
-                   axom::fmt::format("round_{}_r{}_end.json", i, m_rank));
-
-          SLIC_ASSERT_MSG(
-            conduit::blueprint::mcarray::is_interleaved(
-              mesh_node["fields/closest_point/values"]),
-            fmt::format("After copy on iteration {}, 'closest_point' field of "
-                        "'mesh_node' is not interleaved",
-                        i));
+          if(m_isVerbose)
+          {
+            dumpNode(queryMesh,
+                     axom::fmt::format("round_{}_r{}_end.json", round, m_rank));
+
+            SLIC_ASSERT_MSG(
+              conduit::blueprint::mcarray::is_interleaved(
+                queryMesh["fields/closest_point/values"]),
+              fmt::format(
+                "After copy on iteration {}, 'closest_point' field of "
+                "'queryMesh' is not interleaved",
+                round));
+          }
+          // Take xferNode out of circulation.
+          // It has completed its trip through the ring.
+          xferNodePtr.reset();
         }
-
-        MPI_Barrier(MPI_COMM_WORLD);
-        slic::flushStreams();
-      }
+      }  // Locally process *xferNodePtr
+
+      SLIC_INFO_IF(
+        m_isVerbose,
+        fmt::format("=======  End of round {}/{} =======", round, m_nranks));
+    }  // round loop
+
+    // Complete non-blocking sends.
+    // Cyrus recommends using conduit's wait_all, but that requires a
+    // C array of conduit::relay::mpi::Request.  We have a list, which
```

Comment:
It would work, except if we have to increase the vector capacity.  The current nodes get moved to another location, invalidating the pointers we passed to MPI.  We probably should talk with conduit about more flexible ways to pass in Requests.

Another need for flexibility is that we now use MPI_Testsome to complete requests without blocking.  This reduces memory use significantly over checking the requests at the very end.

Reply:
Thanks @gunney1.

It'd probably be a good idea to investigate how the memory usage scales.